authenticator

Maltyz opened the pull request.

Maltyz committed:
- Bump version
- Add a way to handle list from call
- Add sample

david-fournier-smartnsoft committed:
Start authenticator implementation

Maltyz committed:
WIP Authenticator

Maltyz committed:
Remove cache files

Maltyz committed:
WIP Authenticator

Maltyz committed:
WIP Authenticator

Maltyz committed:
Authenticator done

Maltyz committed:
Bump version

david-fournier-smartnsoft commented on retrofit/src/main/kotlin/com/smartnsoft/ws/retrofit/AuthJacksonRetrofitWebServiceCaller.kt:
```diff
@@ -0,0 +1,209 @@
+      authProvider.apply {
+        val accessToken = getAccessToken()
+
+        // We don't want to try to refresh the token more than one time
```
once

david-fournier-smartnsoft commented on retrofit/src/main/kotlin/com/smartnsoft/ws/retrofit/AuthJacksonRetrofitWebServiceCaller.kt:
```diff
@@ -0,0 +1,209 @@
+          {
+            setAccessToken(executeAuth(authService?.refreshToken(getAuthRoute(), accessToken.refreshToken)))
+          }
+          catch (exception: Exception)
```
Possibilité de réduire le scope ?

david-fournier-smartnsoft commented on retrofit/src/main/kotlin/com/smartnsoft/ws/retrofit/AuthJacksonRetrofitWebServiceCaller.kt:
```diff
@@ -0,0 +1,209 @@
+          getAccessToken()?.apply {
+            val newAuthorization = "${this.tokenType} ${this.accessToken}"
+
+            // We don't want to try the call two times with the same token
```
twice

david-fournier-smartnsoft commented on retrofit/src/main/kotlin/com/smartnsoft/ws/retrofit/AuthJacksonRetrofitWebServiceCaller.kt:
```diff
@@ -0,0 +1,209 @@
+          }
+        }
+
+        setAccessToken(null)
```
Mettre un inline comment pour dire que c'est le cas de fail

david-fournier-smartnsoft commented on retrofit/src/main/kotlin/com/smartnsoft/ws/retrofit/AuthJacksonRetrofitWebServiceCaller.kt:
```diff
@@ -0,0 +1,209 @@
+      val newRequest = chain.request().newBuilder()
+
+      authProvider.apply {
+        getXApiKey().takeIf { XApiKey -> XApiKey.isNotBlank() }?.apply {
```
ajouter un test réseau sans api key

david-fournier-smartnsoft commented on retrofit/src/main/kotlin/com/smartnsoft/ws/retrofit/AuthJacksonRetrofitWebServiceCaller.kt:
```diff
@@ -0,0 +1,209 @@
+      var responseHolder: Response? = response
+      var result = 1
+
+      while (responseHolder?.priorResponse() != null)
```
ajouter un paramètre à la classe qui donne le MAX_RETRIES et l'ajouter dans la boucle while

ludovicroland commented on retrofit/src/main/kotlin/com/smartnsoft/ws/retrofit/AccessToken.kt:
```diff
@@ -0,0 +1,22 @@
+import java.io.Serializable
+
+/**
+ * The class description here.
```
@Maltyz : great description!

Maltyz replied:
C DAVID MDR

Maltyz replied:
FIXED

Maltyz replied:
thx

n0tnull replied:
Add a proper description or remove this line

Maltyz replied:
done

david-fournier-smartnsoft commented on retrofit/src/main/kotlin/com/smartnsoft/ws/retrofit/AuthJacksonRetrofitWebServiceCaller.kt:
```diff
@@ -0,0 +1,209 @@
+    }
+  }
+
+  private val httpAuthClient: OkHttpClient by lazy {
```
développer la fonction

david-fournier-smartnsoft reviewed (changes requested):
Faire les quelques remarques

Maltyz committed:
- DeBump version a bit
- Apply all changes from PR

Maltyz committed:
- Add execute with error option
- Add error return for login
- Add Throws annotation

Maltyz reviewed (commented):
Nouvelle PR

n0tnull commented on build.gradle:
```diff
@@ -11,7 +11,7 @@ buildscript
   dependencies
   {
-    classpath 'com.android.tools.build:gradle:3.3.2'
+    classpath 'com.android.tools.build:gradle:3.4.1'
```
Latest version is 3.4.2 ;)

Maltyz replied:
updated

n0tnull commented on retrofit/src/main/kotlin/com/smartnsoft/ws/retrofit/AuthJacksonRetrofitWebServiceCaller.kt:
```diff
@@ -0,0 +1,225 @@
+  companion object
+  {
+
+    const val MAX_RETRIES = 1
```
Why is this constant public and not protected ?

Maltyz replied:
done

n0tnull commented on retrofit/src/main/kotlin/com/smartnsoft/ws/retrofit/AuthJacksonRetrofitWebServiceCaller.kt:
```diff
@@ -0,0 +1,225 @@
+            setAccessToken(null)
+          }
+
+          getAccessToken()?.apply {
```
It could be better to keep the result of the refreshToken method. Then you would be able to refactor the call to setAccessToken but also the apply scope

Maltyz replied:
done

n0tnull commented on retrofit/src/main/kotlin/com/smartnsoft/ws/retrofit/AuthJacksonRetrofitWebServiceCaller.kt:
```diff
@@ -0,0 +1,225 @@
+   * @return the success of the connection.
+   */
+  @WorkerThread
+  protected fun loginUser(username: String, password: String): ResponseWithError<AccessToken, ErrorResponse>?
```
Here is a smaller function body doing the same thing : 
 
```suggestion
return try
    {
      executeAuth(authService?.authToken(authProvider.getAuthRoute(), username, password))
    }
    catch (exception: IOException)
    {
      warn("Call of login failed with exception: ${exception.printStackTrace()}")
      null
    }
    catch (exception: JacksonExceptions.JacksonParsingException)
    {
      warn("Call of login failed with exception: ${exception.printStackTrace()}")
      null
    }.also {
      authProvider.setAccessToken(it?.successResponse)
    }
```

Maltyz replied:
Done

n0tnull commented on retrofit/src/main/kotlin/com/smartnsoft/ws/retrofit/AuthJacksonRetrofitWebServiceCaller.kt:
```diff
@@ -0,0 +1,225 @@
+
+  @WorkerThread
+  @Throws(IOException::class, JacksonExceptions.JacksonParsingException::class)
+  private fun executeAuth(call: Call<AccessToken>?): ResponseWithError<AccessToken, ErrorResponse>?
```
Another suggestion : 
```suggestion
return call?.request()?.let { request ->
      debug("Starting execution of auth call ${request.method()} to ${request.url()}")

      val newRequest = request.newBuilder().build()
      val response: Response? = httpAuthClient.newCall(newRequest).execute()
      val success = response?.isSuccessful
      val responseBody = response?.body()?.string()
      
      return@let ResponseWithError<AccessToken, ErrorResponse>().apply {
        if (success == true)
        {
          successResponse = mapResponseToObject(responseBody, AccessToken::class.java)
        }
        else
        {
          errorResponse = mapResponseToObject(responseBody, ErrorResponse::class.java)
        }
      }
    }
```

n0tnull commented on retrofit/src/main/kotlin/com/smartnsoft/ws/retrofit/CacheException.kt:
```diff
@@ -0,0 +1,10 @@
+ * @author David Fournier
+ * @since 2019.06.19
+ */
+class CacheException(message: String?) : RuntimeException(message)
```
Why is this inheriting from RuntimeException ? 🤔

Maltyz replied:
it's ok

n0tnull commented on retrofit/src/main/kotlin/com/smartnsoft/ws/retrofit/ErrorResponse.kt:
```diff
@@ -0,0 +1,19 @@
+data class ErrorResponse
+(
+    @JsonProperty("status_code") var statusCode: Int,
+    @JsonProperty("message") var message: String
```
Why a var and not a val ?

Maltyz replied:
because i'm dumb

n0tnull commented on retrofit/src/main/kotlin/com/smartnsoft/ws/retrofit/RetrofitWebServiceCaller.kt:
```diff
@@ -591,7 +637,78 @@ abstract class RetrofitWebServiceCaller<out API>(api: Class<API>,
+        responseWithError.errorResponse = mapResponseToObject(responseBody, errorClazz)
+      }
+
+      return responseWithError
```
Same suggestion as before

n0tnull commented on retrofitsample/src/main/kotlin/com/smartnsoft/retrofitsample/ws/InfoAPI.kt:
```diff
@@ -0,0 +1,45 @@
+import java.io.Serializable
+
+/**
+ * The class description here.
```
:)

n0tnull commented on retrofitsample/src/test/java/test/Authenticator.kt:
```diff
@@ -0,0 +1,360 @@
+                  else
+                    return MockResponse().setResponseCode(401)
+                        .setBody("{\n" +
+                            "  \"status_code\": 401,\n" +
```
For a better readability of the JSON, you can use the 3 quotes system """ """

```suggestion
"""{
                    "status_code": 403,
                    "message": "Forbidden"
                    }"""
```

Maltyz replied:
refactored

n0tnull commented on retrofitsample/src/test/java/test/SetCache.kt:
```diff
@@ -0,0 +1,65 @@
+import java.io.File
+
+/**
+ * The class description here.
```
Same thing :)

Maltyz committed:
- Bump version
- Update dependencies
- Do some refactoring (PR)
- Remove droid4me dependencies on RetrofitWebServiceCaller
- Add a common module for droid4me classes we need (logger / exceptions)
- Organize classes in packages

Maltyz committed:
- Remove common package
- Change visibility of properties
- Change the way we build the client

Maltyz committed:
- Split login and refresh in 2 routes
- Change login and refresh signature method
- Add a way to change login and refresh endpoint in the authProvider
- ResponseWithError is now a DataClass
- Update Authenticator tests

Maltyz committed:
- Refresh is now working
- Small changes on auth bo
- We can now consume response multiple times with peekBody
- Add logs on refresh error
- Add real server test classes + test class

Maltyz committed:
- Change the way we handle exceptions

Maltyz committed:
- Update one assert

n0tnull commented on build.gradle:
```diff
@@ -2,7 +2,7 @@
 buildscript
 {
-  ext.kotlin_version = '1.3.20'
+  ext.kotlin_version = '1.3.41'
```
1.3.50 has been released, maybe we could update :)

n0tnull commented on build.gradle:
```diff
@@ -11,7 +11,7 @@ buildscript
   dependencies
   {
-    classpath 'com.android.tools.build:gradle:3.4.1'
+    classpath 'com.android.tools.build:gradle:3.4.2'
```
Same thing with version 3.5.0

n0tnull commented on common/build.gradle:
```diff
@@ -0,0 +1,32 @@
+  api("com.android.support:appcompat-v7:27.1.1")
+}
+
+repositories {
```
Why do you have a "repositories" node here ? 🤔

Maltyz replied:
deleted module, forget to remove all the files

n0tnull commented on retrofit/build.gradle:
```diff
@@ -34,18 +34,15 @@ android
-  api ("com.smartnsoft:droid4me:2.8.0")
-  implementation ("com.smartnsoft:droid4me.ext:2.7.0")
   implementation ("org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version")
+  api("com.android.support:appcompat-v7:27.1.1")
```
You should update to appcompat 28.0.0. If not, it might cause some compatibility issues

n0tnull commented on retrofit/src/main/kotlin/com/smartnsoft/ws/exception/CallException.kt:
```diff
@@ -0,0 +1,67 @@
+  : Exception(message, throwable)
+{
+
+  constructor(throwable: Throwable? = null) : this(null, throwable, 0)
```
Since you have @JvmOverloads on the main constructor with default values, is it really useful to have these 3 additional constructors ?

n0tnull commented on retrofit/src/main/kotlin/com/smartnsoft/ws/exception/CallException.kt:
```diff
@@ -0,0 +1,67 @@
+  companion object
+  {
+
+    private val serialVersionUID = 4869741128441615773L
```
Is it a legacy value from Java ? Do we still need it ?

n0tnull commented on retrofit/src/main/kotlin/com/smartnsoft/ws/exception/JacksonExceptions.kt:
```diff
@@ -0,0 +1,69 @@
+    companion object
+    {
+
+      private val serialVersionUID = 1L
```
Same thing for the two serialVersionIUID in this file

n0tnull commented on retrofit/src/main/kotlin/com/smartnsoft/ws/exception/JacksonExceptions.kt:
```diff
@@ -0,0 +1,69 @@
+    : CallException(message, throwable, statusCode)
+  {
+
+    constructor(throwable: Throwable? = null) : this(null, throwable, 0)
```
Why all these constructors again ?

n0tnull commented on common/build.gradle:
```diff
@@ -0,0 +1,32 @@
+
+dependencies {
+  implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
+  api("com.android.support:appcompat-v7:27.1.1")
```
You could have issues with appcompat version 27 when compiling for API 28 !

Maltyz replied:
deleted

n0tnull commented on retrofit/src/main/kotlin/com/smartnsoft/ws/retrofit/bo/LoginBody.kt:
```diff
@@ -0,0 +1,11 @@
+ */
+data class LoginBody
+(
+    val username: String,
```
We should have Jackson annotation here. It'll be easier to understand what's going on and avoid possible issues with proguard

n0tnull commented on retrofitsample/build.gradle:
```diff
@@ -43,13 +43,19 @@ android
-  implementation ("com.android.support:appcompat-v7:27.1.0")
-  implementation ("com.android.support.constraint:constraint-layout:1.1.3")
+  implementation("org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version")
+  implementation("com.android.support:appcompat-v7:27.1.0")
```
Update to 28.0.0

Maltyz committed:
- Run tests with multiple parameters

Maltyz committed:
- Update dependencies
- Fixes after review
- Remove unused files

Maltyz committed:
- Update all fields for business objects
